File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/admin/products/ProductsAdminScreen.kt
// presentation/ui/admin/products/ProductsAdminScreen.kt
package com.shopapp.presentation.ui.admin.products

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import coil.compose.AsyncImage
import com.shopapp.domain.model.Product
import com.shopapp.presentation.components.LoadingScreen
import com.shopapp.presentation.components.ErrorScreen
import com.shopapp.presentation.viewmodel.ProductStockFilter
import com.shopapp.presentation.viewmodel.ProductsAdminViewModel
import com.shopapp.theme.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ProductsAdminScreen(
    viewModel: ProductsAdminViewModel = hiltViewModel(),
) {
    val state      by viewModel.state.collectAsState()
    val filtered   by viewModel.filtered.collectAsState()
    val formState  by viewModel.formState.collectAsState()
    val categories by viewModel.categories.collectAsState() // ← desde el ViewModel

    var showForm      by remember { mutableStateOf(false) }
    var editTarget    by remember { mutableStateOf<Product?>(null) }
    var restockTarget by remember { mutableStateOf<Product?>(null) }
    var deleteTarget  by remember { mutableStateOf<Product?>(null) }
    var snackMsg      by remember { mutableStateOf<String?>(null) }

    val snackbarHostState = remember { SnackbarHostState() }

    LaunchedEffect(snackMsg) {
        snackMsg?.let {
            snackbarHostState.showSnackbar(it)
            snackMsg = null
        }
    }

    Scaffold(
        snackbarHost   = { SnackbarHost(snackbarHostState) },
        containerColor = Background,
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .background(Background),
        ) {
            // ── Header ────────────────────────────────────────────
            Surface(color = Surface, tonalElevation = 0.dp) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Row(
                        modifier              = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment     = Alignment.CenterVertically,
                    ) {
                        Column {
                            Text(
                                "Productos",
                                style      = MaterialTheme.typography.headlineMedium,
                                fontWeight = FontWeight.Bold,
                                color      = TextPrimary,
                            )
                            Text(
                                "${state.total} productos",
                                style = MaterialTheme.typography.bodySmall,
                                color = TextSecondary,
                            )
                        }
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            IconButton(onClick = viewModel::load) {
                                Icon(Icons.Default.Refresh, null, tint = TextSecondary)
                            }
                            Button(
                                onClick = { editTarget = null; showForm = true },
                                colors  = ButtonDefaults.buttonColors(
                                    containerColor = Accent,
                                    contentColor   = AccentOnDark,
                                ),
                                shape          = MaterialTheme.shapes.medium,
                                contentPadding = PaddingValues(horizontal = 16.dp, vertical = 10.dp),
                            ) {
                                Icon(Icons.Default.Add, null, modifier = Modifier.size(18.dp))
                                Spacer(Modifier.width(4.dp))
                                Text("Nuevo", fontWeight = FontWeight.Bold)
                            }
                        }
                    }

                    Spacer(Modifier.height(12.dp))

                    OutlinedTextField(
                        value         = state.search,
                        onValueChange = viewModel::setSearch,
                        placeholder   = { Text("Buscar producto...", color = TextFaint) },
                        leadingIcon   = { Icon(Icons.Default.Search, null, tint = TextSecondary) },
                        singleLine    = true,
                        modifier      = Modifier.fillMaxWidth(),
                        shape         = MaterialTheme.shapes.medium,
                        colors        = OutlinedTextFieldDefaults.colors(
                            focusedBorderColor   = Accent,
                            unfocusedBorderColor = Border,
                            cursorColor          = Accent,
                        ),
                    )

                    Spacer(Modifier.height(10.dp))

                    LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        items(ProductStockFilter.entries) { filter ->
                            FilterChip(
                                selected = state.stockFilter == filter,
                                onClick  = { viewModel.setStockFilter(filter) },
                                label    = {
                                    Text(filter.label, style = MaterialTheme.typography.labelSmall)
                                },
                                colors = FilterChipDefaults.filterChipColors(
                                    selectedContainerColor = Accent,
                                    selectedLabelColor     = AccentOnDark,
                                    containerColor         = Surface2,
                                    labelColor             = TextSecondary,
                                ),
                            )
                        }
                    }
                }
            }

            // ── Lista ──────────────────────────────────────────────
            when {
                state.isLoading     -> LoadingScreen("Cargando productos...")
                state.error != null -> ErrorScreen(state.error!!, onRetry = viewModel::load)
                filtered.isEmpty()  -> {
                    Box(Modifier.fillMaxSize(), Alignment.Center) {
                        Column(horizontalAlignment = Alignment.CenterHorizontally) {
                            Text("📦", fontSize = 48.sp)
                            Spacer(Modifier.height(12.dp))
                            Text(
                                if (state.search.isBlank() && state.stockFilter == ProductStockFilter.ALL)
                                    "Sin productos"
                                else
                                    "Sin resultados",
                                style      = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.Bold,
                                color      = TextPrimary,
                            )
                        }
                    }
                }
                else -> {
                    LazyColumn(
                        modifier            = Modifier.fillMaxSize(),
                        contentPadding      = PaddingValues(16.dp),
                        verticalArrangement = Arrangement.spacedBy(10.dp),
                    ) {
                        items(filtered, key = { it.id }) { product ->
                            ProductAdminCard(
                                product   = product,
                                onToggle  = { viewModel.toggleActive(product.id, !product.isActive) },
                                onEdit    = { editTarget = product; showForm = true },
                                onRestock = { restockTarget = product },
                                onDelete  = { deleteTarget = product },
                            )
                        }
                    }
                }
            }
        }
    }

    // ── Bottom Sheet: formulario crear / editar ────────────────────
    if (showForm) {
        ProductFormSheet(
            initial    = editTarget,
            categories = categories,
            formState  = formState,
            onSave     = { payload ->
                if (editTarget != null) viewModel.updateProduct(editTarget!!.id, payload)
                else viewModel.createProduct(payload)
            },
            onDismiss = {
                showForm   = false
                editTarget = null
                viewModel.resetFormState()
            },
        )
    }

    // ── Dialog: restock ───────────────────────────────────────────
    restockTarget?.let { product ->
        RestockDialog(
            product   = product,
            onRestock = { qty ->
                viewModel.restock(product.id, qty) { msg ->
                    snackMsg      = msg
                    restockTarget = null
                }
            },
            onDismiss = { restockTarget = null },
        )
    }

    // ── Dialog: confirmación de eliminación ───────────────────────
    deleteTarget?.let { product ->
        AlertDialog(
            onDismissRequest = { deleteTarget = null },
            containerColor   = Surface,
            shape            = MaterialTheme.shapes.large,
            title            = { Text("¿Eliminar producto?", color = TextPrimary) },
            text             = {
                Text(
                    "\"${product.name}\" se eliminará permanentemente.",
                    color = TextSecondary,
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    viewModel.deleteProduct(product.id)
                    deleteTarget = null
                }) {
                    Text("Eliminar", color = Error, fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { deleteTarget = null }) {
                    Text("Cancelar", color = TextSecondary)
                }
            },
        )
    }
}

// ── ProductAdminCard ──────────────────────────────────────────────────────────

@Composable
private fun ProductAdminCard(
    product:   Product,
    onToggle:  () -> Unit,
    onEdit:    () -> Unit,
    onRestock: () -> Unit,
    onDelete:  () -> Unit,
) {
    Surface(
        shape    = MaterialTheme.shapes.large,
        color    = if (product.isActive) Surface else Surface.copy(alpha = 0.6f),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier          = Modifier.padding(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Imagen
            Box(
                modifier         = Modifier
                    .size(56.dp)
                    .background(Surface2, MaterialTheme.shapes.medium),
                contentAlignment = Alignment.Center,
            ) {
                if (product.imageUrl != null) {
                    AsyncImage(
                        model              = product.imageUrl,
                        contentDescription = product.name,
                        contentScale       = ContentScale.Crop,
                        modifier           = Modifier.fillMaxSize(),
                    )
                } else {
                    Text("📦", fontSize = 22.sp)
                }
            }

            Spacer(Modifier.width(12.dp))

            // Info
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text       = product.name,
                    style      = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    color      = TextPrimary,
                    maxLines   = 1,
                    overflow   = TextOverflow.Ellipsis,
                )
                Text(
                    text  = product.categoryName ?: "Sin categoría",
                    style = MaterialTheme.typography.bodySmall,
                    color = TextSecondary,
                )
                Row(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment     = Alignment.CenterVertically,
                ) {
                    Text(
                        "$${"%.2f".format(product.price)}",
                        style      = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Bold,
                        color      = Accent,
                    )
                    // Badge de stock
                    Surface(
                        color = when {
                            product.stock == 0 -> Error.copy(alpha = 0.15f)
                            product.stock < 5  -> Warning.copy(alpha = 0.15f)
                            else               -> Success.copy(alpha = 0.12f)
                        },
                        shape = MaterialTheme.shapes.extraSmall,
                    ) {
                        Text(
                            text = when {
                                product.stock == 0 -> "Agotado"
                                else               -> "${product.stock} uds."
                            },
                            color = when {
                                product.stock == 0 -> Error
                                product.stock < 5  -> Warning
                                else               -> Success
                            },
                            fontSize   = 10.sp,
                            fontWeight = FontWeight.Bold,
                            modifier   = Modifier.padding(horizontal = 6.dp, vertical = 2.dp),
                        )
                    }
                }
            }

            // Acciones
            Column(horizontalAlignment = Alignment.End) {
                Switch(
                    checked         = product.isActive,
                    onCheckedChange = { onToggle() },
                    colors          = SwitchDefaults.colors(
                        checkedThumbColor    = AccentOnDark,
                        checkedTrackColor    = Accent,
                        uncheckedTrackColor  = Surface2,
                        uncheckedBorderColor = Border,
                    ),
                    modifier = Modifier.size(width = 48.dp, height = 28.dp),
                )
                Spacer(Modifier.height(4.dp))
                Row {
                    IconButton(onClick = onRestock, modifier = Modifier.size(32.dp)) {
                        Icon(
                            Icons.Default.Inventory, null,
                            tint     = Accent,
                            modifier = Modifier.size(18.dp),
                        )
                    }
                    IconButton(onClick = onEdit, modifier = Modifier.size(32.dp)) {
                        Icon(
                            Icons.Default.Edit, null,
                            tint     = TextSecondary,
                            modifier = Modifier.size(18.dp),
                        )
                    }
                    IconButton(onClick = onDelete, modifier = Modifier.size(32.dp)) {
                        Icon(
                            Icons.Default.Delete, null,
                            tint     = Error,
                            modifier = Modifier.size(18.dp),
                        )
                    }
                }
            }
        }
    }
}

File: modulo4/flutter_shop_app/lib/presentation/screens/admin/products_admin_screen.dart
// lib/presentation/screens/admin/products_admin_screen.dart

import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../theme/app_colors.dart';
import '../../../core/utils/formatters.dart';
import '../../../data/repository/category_repository_impl.dart';
import '../../../domain/model/category.dart';
import '../../../domain/model/product.dart';
import '../../providers/products_admin_provider.dart';
import '../../widgets/product_form.dart';
import '../../widgets/restock_dialog.dart';

class ProductsAdminScreen extends ConsumerStatefulWidget {
  const ProductsAdminScreen({super.key});

  @override
  ConsumerState<ProductsAdminScreen> createState() => _ProductsAdminScreenState();
}

class _ProductsAdminScreenState extends ConsumerState<ProductsAdminScreen> {
  List<Category> _categories = [];

  @override
  void initState() {
    super.initState();
    ref.read(categoryRepositoryProvider).getCategories().then((cats) {
      if (mounted) setState(() => _categories = cats);
    });
  }

  @override
  Widget build(BuildContext context) {
    final state    = ref.watch(productsAdminProvider);
    final filtered = state.filtered;

    return Column(
      children: [
        // ── Header ──────────────────────────────────────────
        Container(
          color:   AppColors.surface,
          padding: const EdgeInsets.fromLTRB(16, 16, 16, 0),
          child:   Column(
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      const Text('Productos',
                          style: TextStyle(
                            color: AppColors.textPrimary, fontSize: 22, fontWeight: FontWeight.bold,
                          )),
                      Text('${state.total} productos',
                          style: const TextStyle(color: AppColors.textSecondary, fontSize: 13)),
                    ],
                  ),
                  Row(
                    children: [
                      IconButton(
                        onPressed: () => ref.read(productsAdminProvider.notifier).load(),
                        icon:      const Icon(Icons.refresh_rounded, color: AppColors.textSecondary),
                      ),
                      ElevatedButton.icon(
                        onPressed: () => showProductForm(context, ref, categories: _categories),
                        icon:      const Icon(Icons.add, size: 18),
                        label:     const Text('Nuevo'),
                        style:     ElevatedButton.styleFrom(
                          minimumSize: const Size(0, 40),
                          padding:     const EdgeInsets.symmetric(horizontal: 16),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
              const SizedBox(height: 12),

              // Búsqueda
              TextField(
                onChanged:  ref.read(productsAdminProvider.notifier).setSearch,
                decoration: const InputDecoration(
                  hintText:   'Buscar producto...',
                  prefixIcon: Icon(Icons.search_rounded, color: AppColors.textSecondary),
                  contentPadding: EdgeInsets.symmetric(vertical: 10),
                ),
                style: const TextStyle(color: AppColors.textPrimary),
              ),
              const SizedBox(height: 10),

              // Chips de filtro
              SizedBox(
                height: 34,
                child:  ListView(
                  scrollDirection: Axis.horizontal,
                  children: ProductStockFilter.values.map((f) => Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child:   ChoiceChip(
                      label:     Text(f.label),
                      selected:  state.stockFilter == f,
                      onSelected:(_) =>
                          ref.read(productsAdminProvider.notifier).setStockFilter(f),
                    ),
                  )).toList(),
                ),
              ),
              const SizedBox(height: 12),
            ],
          ),
        ),

        // ── Lista ─────────────────────────────────────────────
        Expanded(
          child: Builder(builder: (_) {
            if (state.isLoading) {
              return const Center(
                child: CircularProgressIndicator(color: AppColors.accent),
              );
            }
            if (state.error != null) {
              return Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text(state.error!, style: const TextStyle(color: AppColors.error)),
                    const SizedBox(height: 12),
                    ElevatedButton(
                      onPressed: () => ref.read(productsAdminProvider.notifier).load(),
                      child:     const Text('Reintentar'),
                    ),
                  ],
                ),
              );
            }
            if (filtered.isEmpty) {
              return const Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text('📦', style: TextStyle(fontSize: 48)),
                    SizedBox(height: 12),
                    Text('Sin productos',
                        style: TextStyle(
                          color: AppColors.textPrimary, fontSize: 18, fontWeight: FontWeight.bold,
                        )),
                  ],
                ),
              );
            }

            return ListView.separated(
              padding:         const EdgeInsets.all(16),
              itemCount:       filtered.length,
              separatorBuilder:(_, __) => const SizedBox(height: 10),
              itemBuilder: (_, i) => _ProductAdminCard(
                product:    filtered[i],
                onToggle:   () => ref.read(productsAdminProvider.notifier)
                    .toggleActive(filtered[i].id, !filtered[i].isActive),
                onEdit:     () => showProductForm(
                  context, ref, initial: filtered[i], categories: _categories,
                ),
                onRestock:  () async {
                  final qty = await showRestockDialog(context, filtered[i]);
                  if (qty != null && context.mounted) {
                    final newStock = await ref
                        .read(productsAdminProvider.notifier)
                        .restock(filtered[i].id, qty);
                    if (context.mounted) {
                      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                        content: Text(
                          newStock != null
                              ? '✅ Stock actualizado: $newStock unidades'
                              : '❌ Error al actualizar el stock',
                        ),
                        backgroundColor:
                            newStock != null ? AppColors.success : AppColors.error,
                      ));
                    }
                  }
                },
                onDelete:   () => _confirmDelete(context, ref, filtered[i]),
              ),
            );
          }),
        ),
      ],
    );
  }

  void _confirmDelete(BuildContext context, WidgetRef ref, Product product) {
    showDialog(
      context: context,
      builder: (_) => AlertDialog(
        backgroundColor: AppColors.surface,
        shape:           RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        title:  const Text('¿Eliminar producto?',
            style: TextStyle(color: AppColors.textPrimary)),
        content:Text('"${product.name}" se eliminará permanentemente.',
            style: const TextStyle(color: AppColors.textSecondary)),
        actions:[
          TextButton(
            onPressed: () => Navigator.pop(context),
            child:     const Text('Cancelar'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ref.read(productsAdminProvider.notifier).deleteProduct(product.id);
            },
            child: const Text('Eliminar',
                style: TextStyle(color: AppColors.error, fontWeight: FontWeight.bold)),
          ),
        ],
      ),
    );
  }
}

// ── ProductAdminCard ──────────────────────────────────────────

class _ProductAdminCard extends StatelessWidget {
  final Product      product;
  final VoidCallback onToggle;
  final VoidCallback onEdit;
  final VoidCallback onRestock;
  final VoidCallback onDelete;

  const _ProductAdminCard({
    required this.product,
    required this.onToggle,
    required this.onEdit,
    required this.onRestock,
    required this.onDelete,
  });

  Color _stockColor() {
    if (product.stock == 0) return AppColors.error;
    if (product.stock < 5)  return AppColors.warning;
    return AppColors.success;
  }

  @override
  Widget build(BuildContext context) => Opacity(
    opacity: product.isActive ? 1.0 : 0.55,
    child:   Container(
      padding:    const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color:        AppColors.surface,
        borderRadius: BorderRadius.circular(14),
        border:       Border.all(color: AppColors.border),
      ),
      child: Row(
        children: [
          // Imagen
          ClipRRect(
            borderRadius: BorderRadius.circular(10),
            child: SizedBox(
              width: 54, height: 54,
              child: product.imageUrl != null
                  ? CachedNetworkImage(
                      imageUrl: product.imageUrl!,
                      fit:      BoxFit.cover,
                      errorWidget: (_, __, ___) => Container(
                        color: AppColors.surface2,
                        child: const Center(child: Text('📦')),
                      ),
                    )
                  : Container(
                      color: AppColors.surface2,
                      child: const Center(
                        child: Text('📦', style: TextStyle(fontSize: 22)),
                      ),
                    ),
            ),
          ),
          const SizedBox(width: 12),

          // Info
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  product.name,
                  style: const TextStyle(
                    color: AppColors.textPrimary, fontWeight: FontWeight.w600,
                  ),
                  maxLines: 1, overflow: TextOverflow.ellipsis,
                ),
                if (product.category != null)
                  Text(product.category!.name,
                      style: const TextStyle(color: AppColors.textSecondary, fontSize: 12)),
                Row(
                  children: [
                    Text(
                      formatPrice(product.price),
                      style: const TextStyle(
                        color: AppColors.accent, fontWeight: FontWeight.bold, fontSize: 14,
                      ),
                    ),
                    const SizedBox(width: 8),
                    Container(
                      padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                      decoration: BoxDecoration(
                        color:        _stockColor().withValues(alpha: 0.12),
                        borderRadius: BorderRadius.circular(6),
                      ),
                      child: Text(
                        product.stock == 0 ? 'Agotado' : '${product.stock} uds.',
                        style: TextStyle(
                          color:      _stockColor(),
                          fontSize:   10,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),

          // Acciones
          Column(
            children: [
              Switch(
                value:       product.isActive,
                onChanged:   (_) => onToggle(),
                activeThumbColor: AppColors.accent,
              ),
              Row(
                children: [
                  _ActionIcon(icon: Icons.inventory_2_outlined, color: AppColors.accent,    onTap: onRestock),
                  _ActionIcon(icon: Icons.edit_outlined,         color: AppColors.textSecondary, onTap: onEdit),
                  _ActionIcon(icon: Icons.delete_outline,        color: AppColors.error,    onTap: onDelete),
                ],
              ),
            ],
          ),
        ],
      ),
    ),
  );
}

class _ActionIcon extends StatelessWidget {
  final IconData     icon;
  final Color        color;
  final VoidCallback onTap;
  const _ActionIcon({required this.icon, required this.color, required this.onTap});

  @override
  Widget build(BuildContext context) => GestureDetector(
    onTap: onTap,
    child: Padding(
      padding: const EdgeInsets.all(4),
      child:   Icon(icon, color: color, size: 20),
    ),
  );
}